nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchRecord.java

```diff
@@ -0,0 +1,354 @@
+                : null;
+        final String typePath = context.getProperty(TYPE_RECORD_PATH).isSet()
+                ? context.getProperty(TYPE_RECORD_PATH).evaluateAttributeExpressions(input).getValue()
+                : null;
+
+        RecordPath path = idPath != null ? recordPathCache.getCompiled(idPath) : null;
+        RecordPath iPath = indexPath != null ? recordPathCache.getCompiled(indexPath) : null;
+        RecordPath tPath = typePath != null ? recordPathCache.getCompiled(typePath) : null;
+
+        int batchSize = context.getProperty(BATCH_SIZE).evaluateAttributeExpressions(input).asInteger();
+        List<FlowFile> badRecords = new ArrayList<>();
+
+        try (final InputStream inStream = session.read(input);
+             final RecordReader reader = readerFactory.createRecordReader(input, inStream, getLogger())) {
+            Record record;
+            List<IndexOperationRequest> operationList = new ArrayList<>();
+            List<Record> originals = new ArrayList<>();
+
+            while ((record = reader.nextRecord()) != null) {
+                final String idx = getFromRecordPath(record, iPath, index);
+                final String t   = getFromRecordPath(record, tPath, type);
+                final IndexOperationRequest.Operation o = IndexOperationRequest.Operation.Index;
+                final String id  = path != null ? getFromRecordPath(record, path, null) : null;
+
+                Map<String, Object> contentMap = (Map<String, Object>) DataTypeUtils.convertRecordFieldtoObject(record, RecordFieldType.RECORD.getRecordDataType(record.getSchema()));
+
+                removeEmpty(contentMap);
+
+                operationList.add(new IndexOperationRequest(idx, t, id, contentMap, o));
+                originals.add(record);
+
+                if (operationList.size() == batchSize) {
+                    BulkOperation bundle = new BulkOperation(operationList, originals, reader.getSchema());
+                    FlowFile bad = indexDocuments(bundle, session, input);
+                    if (bad != null) {
+                        badRecords.add(bad);
+                    }
+
+                    operationList.clear();
+                    originals.clear();
+                }
+            }
+
+            if (operationList.size() > 0) {
+                BulkOperation bundle = new BulkOperation(operationList, originals, reader.getSchema());
+                FlowFile bad = indexDocuments(bundle, session, input);
+                if (bad != null) {
+                    badRecords.add(bad);
+                }
+            }
+
+            session.transfer(input, REL_SUCCESS);
+        } catch (ElasticsearchError ese) {
+            String msg = String.format("Encountered a server-side problem with Elasticsearch. %s",
+                    ese.isElastic() ? "Moving to retry." : "Moving to failure");
+            getLogger().error(msg, ese);
+            Relationship rel = ese.isElastic() ? REL_RETRY : REL_FAILURE;
```

Comment:
Should retry flowfiles be penalized? I think that's the common pattern for external (but retryable) issues